### PR TITLE
 Add Effects: BulgePinch, Glitch, RadialBlur and Twist

### PR DESCRIPTION
**Pull request by Quarkstar**
Add BulgePinch, Glitch, RadialBlur and Twist

### BulgePinch
Screenshot from 2020-03-14 12-48-14
### Glitch
Screenshot from 2020-03-14 12-49-13
### RadialBlur
Screenshot from 2020-03-14 13-08-23
### Twist
Screenshot from 2020-03-14 12-53-22
Twist has a parameter named `padding`, and I don't find any change after changing it.(In https://pixijs.io/pixi-filters/tools/demo/ it can't be changed). I added it temporary. Should it be deleted? 

## A problem
I found `blur`'s `kernelSize` cannot be setted >= 11 and I don't know why. Can someone do more tests on it? :)
Screenshot from 2020-03-14 13-56-21
### More information 
I extended `gdjs.PixiFiltersTools.clampValue` and found IDE crashed after I setting kernelSize to 11.(In isometric-game) 
Then I found it also crashed in previous commit. It works when `kernelSize` <= 9.

### CODE REVIEW

File: Extensions/Effects/JsExtension.js [hunk 1]
```diff
@@ -520,6 +552,113 @@ module.exports = {
...
+        )
+    );
+    glitchProperties.set(
+      'average',
+      new gd.PropertyDescriptor(/* defaultValue= */ 'false')
+        .setLabel(_('Average'))
+        .setType('boolean')
+        .setDescription(
+          _(
+            'true will divide the bands roughly based on equal amounts where as setting to false will vary the band sizes dramatically (more random looking)'
```
Comment (4ian) [inline Extensions/Effects/JsExtension.js]: This will be shown as a checkbox in the UI, so the description should not contain "true will [...]". It must describe what happen if you check the checkbox :)

File: Extensions/Effects/JsExtension.js [hunk 2]
```diff
@@ -274,6 +273,39 @@ module.exports = {
...
+      .addEffect('BulgePinch')
+      .setFullName(_('Bulge Pinch'))
+      .setDescription(_('Bulges or pinches the image in a circle.'))
+      .addIncludeFile('Extensions/Effects/pixi-filters/filter-bulge-pinch.js')
+      .addIncludeFile('Extensions/Effects/bulge-pinch-pixi-filter.js');
+    const bulgePinchProperties = bulgePinchEffect.getProperties();
+    bulgePinchProperties.set(
+      'centerX',
+      new gd.PropertyDescriptor(/* defaultValue= */ '0.5')
+        .setLabel(_('Center X (between 0 and 1, 0.5 is window middle)'))
```
Comment (4ian) [inline Extensions/Effects/JsExtension.js]: I know this was done in other filters - but I think we should use center X/Y in pixels, rather than a percentage of the window size.

The reason is that filters will soon work on objects - so there is no "window" size in this context.

**Comment by 4ian** [comment]
Thanks for the PR! Looks mostly good.

@blurymind @Bouh Could you help me by reviewing this PR? :) 

The main thing I'm wondering is if we can remove the window.innerWidth/window.innerHeight and just use values in pixel (for radius, center X, center Y, etc...).
Maybe we could even do this for existing filters - this is a detail I missed previously. 
The reason is that in the future the effects will be usable on objects.

**Comment by Quarkstar** [comment]
Thank you! And you replied so quickly that I haven't sent my last comment :joy:
   
I found `AppVeyor` failed in last commit. I am trying to find out why. Need I remove `window` for other filters? I'd glad to.

**Comment by Quarkstar** [comment]
After removing `window`, I find it is not intuitive enough for the users. I think it is much better if we can add a hint about the height and width of image. I think of two ways.

1. Add height and width of image in description(In `JsExtension.js`). It needs us to pass in the image information as a parameter(may need a lot of changes).
2. Users' input is parameter between 0 and 1,  and we get the image information from layer(or objects) and multiply them together(In `xxx-pixi-filter.js`), just like what `PIXI.filters.BulgePinchFilter.center` does. But I don't know if we can get image information.

Thanks again for your patience!

**Comment by blurymind** [comment]
I think we need to keep this intuitive somehow. The filters need to know the center point of the thing we attach them to. Perhaps we could pass the width/height behind the scenes based on whether its attached to an object or a layer somehow? I dont like the idea of burdening the user with the need to always have to do this manually :)

**Comment by Bouh** [comment]
- Seed is missing for Glitch effect, have you tried to randomize this properties, like said the documentation ?

**Comment by 4ian** [comment]
I suggested this because pixels are maybe actually more intuitive than a 0-1 scale (people could ask "why not 0-100?" - even if 0-1 is better). This could be fixed by a better input in the UI (see the rest of my message)

* Having the radius dependent on the window width is actually counter intuitive (why not the window height? 😬). 
It is also not preventing us to put a proper default  value (something like 200 is good enough to see the effect, whatever your window, or in the future object, size).
* While it's true that 0;0 as the default center X/Y positions is not maybe perfect, the effect will be still visible no? It needs a bit more effort to center the effect if your game resolution is changed. If we want to keep 0-1 because it's more convenient, then we must:
  * Pass the width/height to use as reference to the effect (currently it's always the window width/height, but in the future there will be objects) - might need a bit of refactoring yes.
  * Maybe think about making a special version of the input shown for these "0-1 range" parameters. Like a [slider](https://material-ui.com/components/slider/) followed by a numeric input: 

[-------|----] [0.63]

What do you think?

**Comment by Bouh** [comment]
0-1 are related to UV from shaders.
If we change 0-1 to something else, we need remap these to 0-1.

**Comment by Quarkstar** [comment]
>     * Seed is missing for Glitch effect, have you tried to randomize this properties, like said the documentation ?

Thank you! In Glitch effect, I add a `boolean` parameter named `animated`. And in `glitch-pixi-filter.js`, there is
```javascript
update: function(filter, layer) {
   if (filter.animated) {
      filter.time += layer.getElapsedTime() / 1000;
      filter.seed = Math.random();
   }
 }
```
(I tried to do the same as `CRTFilter`).
Should I add a new parameter named `seed`? @Bouh

**Comment by 4ian** [comment]
> I tried to do the same as CRTFilter

What you've done looks correct to me :)

File: Extensions/Effects/JsExtension.js [hunk 3]
```diff
@@ -520,6 +552,113 @@ module.exports = {
...
+      .addEffect('Glitch')
+      .setFullName(_('Glitch'))
+      .setDescription(_('Applies a glitch effect to an object.'))
+      .addIncludeFile('Extensions/Effects/pixi-filters/filter-glitch.js')
+      .addIncludeFile('Extensions/Effects/glitch-pixi-filter.js');
+    const glitchProperties = glitchEffect.getProperties();
+    glitchProperties.set(
+      'slices',
+      new gd.PropertyDescriptor(/* defaultValue= */ '5')
+        .setLabel(_('Slices (between 2 and 20)'))
```
[the remark this answers is not in the document]
Comment (Quarkstar) [reply Extensions/Effects/JsExtension.js]: Thank you!
I used to find it in (https://pixijs.io/pixi-filters/docs/PIXI.filters.GlitchFilter.html#slices) and there is only a default value. So I set it the same as (https://pixijs.io/pixi-filters/tools/demo/).
Can you tell me where I can check the parameter range?

[the remark this answers is not in the document]
Comment (Bouh) [reply Extensions/Effects/JsExtension.js]: I look [in code](https://pixijs.io/pixi-filters/docs/filters_glitch_src_GlitchFilter.js.html#line35), there is no max limit :) you can tell between 2 and infinite.

**Comment by Bouh** [comment]
> (I tried to do the same as `CRTFilter`).
> Should I add a new parameter named `seed`? @Bouh

I didn't realize there was a solution already in place.
It's good for me too :)

**Comment by blurymind** [comment]
I think we should handle all technical values automatically for the user and leave only the creative ones exposed in the IDE . So seed, animate, width, height - gdevelop should handle all of them in the background and not ask the user to manually increment or set seed to random. Whenever you have animate, have a boolean to enable/disable the animation - which you already have done.

I suggest to figure out the width/height thing when support for object filters is being added. It would be one of the things that need to be figured out anyway

File: Extensions/Effects/JsExtension.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (Quarkstar) [reply Extensions/Effects/JsExtension.js]: Thank you!

**Comment by Quarkstar** [comment]
>  It would be interesting to expose animatin speed to the user.

It looks really interesting and I like it. 
 
>  Maybe think about making a special version of the input shown for these "0-1 range" parameters. Like a slider followed by a numeric input.

I think [Slider with input field](https://material-ui.com/components/slider/#with-input-field) is a good choice! 
1. More intuitive. It's also easy for users to input a specific value.
2. Clearer range of parameters. And we don't have to add `(bewteen x and y)` everywhere.

For this pr, what should I do next?

**Comment by 4ian** [comment]
> So maybe instead of a boolean, animated could be a number.
> 0- paused, 0.5- half speed, 1- normal speed, 2- double the speed?

Yes that looks good.

> For this pr, what should I do next?

The slider won't be for this PR. For this PR:

- Decide if you want to add the speed. If not, we should probably add it in a next PR. As this will remove an option ("animated") to replace it by another ("animationSpeed", or something like this), I would suggest we do it as early as possible.
- Decide if you keep 0-1 for some fields, and explain why.
- Confirm that you've used pixels as much as possible for fields where 0-1 does not make sense (like the radius).

Basically, try to make a quick summary to explain all decisions and make sure we can merge this PR without having to fix things just after :) 

Thanks for your work so far! Appreciate it.

**Comment by blurymind** [comment]
for animationSpeed I suggest having 1 be the default and including a hint about the range with 0 being paused :)  You will need to have a look at the other filters that are animated

**Comment by 4ian** [comment]
> for animationSpeed I suggest having 1 be the default and including a hint about the range with 0 being paused :) 

Yep, sounds fair! That seems intuitive, that's how a time scale works already.

>  You will need to have a look at the other filters that are animated

We can update the other effects in another PR.

**Comment by Quarkstar** [comment]
> Decide if you want to add the speed.

I would like to add it in next pr, because it will change all filters that are animated. I will do it soon.

> Decide if you keep 0-1 for some fields, and explain why.
- In `BulgePinch, Glitch, RadialBlur, Twist`
Only `bulgePinchEffect.center` keeps 0-1. Unlike other filter in `PIXI.Filter`, `x` and `y` in `PIXI.filters.BulgePinchFilter.center` are between 0 and 1. So I don't change it.

- In `OldFlim, ColorReplace, Dot`
In `oldFilmEffect`, `sepia, noise, scratchDensity,  vignettingAlpha, vignettingBlur` are between 0-1. In `colorReplaceEffecr`, `epsilon` is between 0-1. All parameters above represent intensity/sensitivity of the effect.
Only `oldFilmEffect.vignetting` keeps 0-1 because `PIXI.Filter.OldFilmFilter.vignetting` is between 0-1.


If we want to change  `bulgePinchEffect.center` and `oldFilmEffect.vignetting`，we should get objects’/layers‘ width and height.

> Confirm that you've used pixels as much as possible for fields where 0-1 does not make sense (like the radius).

Yes. I have checked it.

**Comment by Quarkstar** [comment]
@4ian 
Can the above summary be accepted? 😄
- I plan to change all filters that are animated in another pr beacuse it is easy to review. (And I will do it as quick as possible.) Should I change it for these four filters right now?
- `BulgePinchFilter.center` and `OldFilmFilter.vignetting` get the height and width of image from `input.filterFrame` directly and they only accept values bewteen 0-1. So I cannot change them into pixels.
```javascript
apply(filterManager, input, output, clear) {
  this.uniforms.dimensions[0] = input.filterFrame.width;
  this.uniforms.dimensions[1] = input.filterFrame.height;
  filterManager.applyFilter(this, input, output, clear);
}
```
Sorry to bother you.

**Comment by 4ian** [comment]
Ok let's go ahead with this then :) 

> I plan to change all filters that are animated in another pr beacuse it is easy to review

Sure, let's do that in another PR.

> BulgePinchFilter.center and OldFilmFilter.vignetting get the height and width of image from input.filterFrame directly and they only accept values bewteen 0-1

Ok that's fair enough.
* Can you verify that the description are properly indicating that this is between 0 and 1? If not, you can make a new PR to fix them.
* We can add a slider + field in a new PR (that's a different task).

Finally, we could probably "fix" the filters already using window.innerWidth/window.innerHeight because ideally we don't want to use that in the filters.

Merging now, thanks a lot for adding these, for your patience and for your investigations :) 👍

**Comment by Quarkstar** [comment]
> Can you verify that the description are properly indicating that this is between 0 and 1? If not, you can make a new PR to fix them.
Yes, they have descriptions.

Thank you!

**Comment by Quarkstar** [comment]
> I think we should handle all technical values automatically for the user and leave only the creative ones exposed in the IDE . So seed, animate, width, height - gdevelop should handle all of them in the background and not ask the user to manually increment or set seed to random. Whenever you have animate, have a boolean to enable/disable the animation - which you already have done. +1
> 
> I suggest to figure out the width/height thing when support for object filters is being added. It would be one of the things that need to be figured out anyway. For layer filters, lets keep what works for now and change it when we have to
> 
> I wonder now if in the future maybe we should expose the animation value to the user. Is there any creative element in that we're not tapping in - like speed of animation or direction. It would be interesting to expose animatin speed to the user - so they can control how fast a filter is animating. So maybe instead of a boolean, animated could be a number.
> 0- paused, 0.5- half speed, 1- normal speed, 2- double the speed?
> 
> So this could be
> 
> ```
> update: function(filter, layer) {
>    if (filter.animationSpeed !== 0) {
>       filter.time += layer.getElapsedTime() / 1000 * filter.animationSpeed;
>       filter.seed = Math.random();
>    }
>  }
> ```

@blurymind @4ian 
I  try it and it doesn't seem to work. :rofl: 
For all filters which have `seed` , I think `Animation Speed` depends on how often we change `filter.seed` (frequency of calling `update` function).  
Because of that, the max value of `animationSpeed` is 1.

So when `animationSpeed` belongs to 0-1, this could be
```javascript
    if (filter.animationSpeed !== 0) {
      filter.time += layer.getElapsedTime() / 1000 * filter.animationSpeed;
      filter.animationCount += filter.animationSpeed;
      if (filter.gdAnimationCount >= 1) {
        filter.seed = Math.random();
        filter.gdAnimationCount = 0
      }
    }
```

- Is my code correct? Is there any way to get animation played at high speed? Or just set `animationSpeed` bewteen 0-1?

**Comment by blurymind** [comment]
So I'm a bit confused. Why do you increment animationCount? Seed shouldn't
affect time. I suppose you have to look at what the specific filter is
expecting to get based on it's demo example at pixi-filters. Unfortunately
I don't think all of them have been designed to take consistently the same
inputs.

**Comment by blurymind** [comment]
I can try to see why it's not working tomorrow.

**Comment by Quarkstar** [comment]
Sorry, that comment is outdated. I change it in https://github.com/4ian/GDevelop/pull/1554.
